[Issue]: Builds with `-DBUILD_WITH_SPARSE=ON` are broken

### Problem Description

Hi,

The changelog for rocm-6.3.0 release mentions:

> Changed rocSPARSE and SuiteSparse to be run-time dependencies by default. The `BUILD_WITH_SPARSE` CMake option can still be used to convert them into build-time dependencies (now off by default).

I tried to turn it on, but build failed:

```
/usr/bin/hipcc -DUSE_PROF_API=1 -D__HIP_PLATFORM_AMD__=1 -Dhipsolver_EXPORTS -DHAVE_ROCSPARSE -I/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0/library/include -I/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0/library/include/internal -I/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0_build/include/hipsolver -I/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0_build/include/hipsolver/internal -I/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0_build/include -I/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0/library/src/include -I/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0/library/src -isystem /usr/include/rocblas  -march=native -O2 -pipe -fdiagnostics-color=always -flto=thin -Werror=odr -Werror=strict-aliasing -std=c++17 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wno-unused-result -MD -MT library/src/CMakeFiles/hipsolver.dir/amd_detail/dlopen/rocsparse.cpp.o -MF library/src/CMakeFiles/hipsolver.dir/amd_detail/dlopen/rocsparse.cpp.o.d -o library/src/CMakeFiles/hipsolver.dir/amd_detail/dlopen/rocsparse.cpp.o -c /var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0/library/src/amd_detail/dlopen/rocsparse.cpp
/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0/library/src/amd_detail/dlopen/rocsparse.cpp:29:1: error: unknown type name 'fp_rocsparse_create_mat_descr'
   29 | fp_rocsparse_create_mat_descr   g_rocsparse_create_mat_descr;
      | ^
/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0/library/src/amd_detail/dlopen/rocsparse.cpp:30:1: error: unknown type name 'fp_rocsparse_destroy_mat_descr'
   30 | fp_rocsparse_destroy_mat_descr  g_rocsparse_destroy_mat_descr;
      | ^
/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0/library/src/amd_detail/dlopen/rocsparse.cpp:31:1: error: unknown type name 'fp_rocsparse_get_mat_type'
   31 | fp_rocsparse_get_mat_type       g_rocsparse_get_mat_type;
      | ^
/var/tmp/portage/sci-libs/hipSOLVER-6.3.0/work/hipSOLVER-rocm-6.3.0/library/src/amd_detail/dlopen/rocsparse.cpp:32:1: error: unknown type name 'fp_rocsparse_get_mat_index_base'
   32 | fp_rocsparse_get_mat_index_base g_rocsparse_get_mat_index_base;
      | ^
```

I think the reason is that you always build `dlopen/cholmod.cpp` and `dlopen/rocsparse.cpp`, even if instead of dlopen user links to cholmod/rocsparse libraries directly.

https://github.com/ROCm/hipSOLVER/blob/747980104cdc22eb184d450c3b44e98cd21b4892/library/src/CMakeLists.txt#L45-L46

### Operating System

Gentoo

### CPU

-

### GPU

-

### ROCm Version

ROCm 6.3.0

### ROCm Component

_No response_

### Steps to Reproduce

_No response_

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

_No response_

### Additional Information

_No response_

--- a/library/src/CMakeLists.txt
+++ b/library/src/CMakeLists.txt
@@ -42,8 +42,6 @@ prepend_path(".." hipsolver_headers_public relative_hipsolver_headers_public)
 
 if(NOT USE_CUDA)
   set(hipsolver_source
-    "${CMAKE_CURRENT_SOURCE_DIR}/amd_detail/dlopen/cholmod.cpp"
-    "${CMAKE_CURRENT_SOURCE_DIR}/amd_detail/dlopen/rocsparse.cpp"
     "${CMAKE_CURRENT_SOURCE_DIR}/amd_detail/hipsolver_conversions.cpp"
     "${CMAKE_CURRENT_SOURCE_DIR}/amd_detail/hipsolver.cpp"
     "${CMAKE_CURRENT_SOURCE_DIR}/amd_detail/hipsolver_dense.cpp"
@@ -52,6 +50,12 @@ if(NOT USE_CUDA)
     "${CMAKE_CURRENT_SOURCE_DIR}/amd_detail/hipsolver_sparse.cpp"
     "${CMAKE_CURRENT_SOURCE_DIR}/common/hipsolver_dense_common.cpp"
   )
+  if(NOT BUILD_WITH_SPARSE)
+    list(APPEND hipsolver_source
+      "${CMAKE_CURRENT_SOURCE_DIR}/amd_detail/dlopen/cholmod.cpp"
+      "${CMAKE_CURRENT_SOURCE_DIR}/amd_detail/dlopen/rocsparse.cpp"
+    )
+  endif()
 else()
   set(hipsolver_source
     "${CMAKE_CURRENT_SOURCE_DIR}/nvidia_detail/hipsolver_conversions.cpp"

